File: ios/FarmaciasDeGuardiaEnSegovia/Services/PDFParsing/ColumnBasedPDFParser.swift
/*
 * Copyright (C) 2025  Bruno Follon (@bFollon)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import Foundation
import PDFKit

/// Base class for parsing PDFs that use a column-based layout.
/// This extracts common functionality that might be shared between different region parsers.
public class ColumnBasedPDFParser {
    /// Represents a column of text in the PDF with its position and content
    struct TextColumn {
        let x: CGFloat
        let width: CGFloat
        var texts: [(y: CGFloat, text: String)]
        
        init(x: CGFloat, width: CGFloat) {
            self.x = x
            self.width = width
            self.texts = []
        }
    }
    
    /// Removes duplicate adjacent text blocks, keeping only the first occurrence
    func removeDuplicateAdjacent(blocks: [(y: CGFloat, text: String)]) -> [(y: CGFloat, text: String)] {
        guard !blocks.isEmpty else { return [] }
        
        var result: [(y: CGFloat, text: String)] = [blocks[0]]
        for i in 1..<blocks.count {
            let current = blocks[i]
            // Only add if text is different from the previous block
            if current.text != result.last?.text {
                result.append(current)
            }
        }
        return result
    }
    
    /// Scans a column in the PDF page and extracts text
    func scanColumn(_ page: PDFPage, column: TextColumn, baseHeight: CGFloat, scanIncrement: CGFloat) -> [(y: CGFloat, text: String)] {
        var texts: [(y: CGFloat, text: String)] = []
        
        // Scan from top to bottom with high precision
        for y in stride(from: 0, to: page.bounds(for: .mediaBox).height, by: scanIncrement) {
            let rect = CGRect(x: column.x, y: y, width: column.width, height: baseHeight)
            if let selection = page.selection(for: rect),
               let text = selection.string?.trimmingCharacters(in: .whitespacesAndNewlines),
               !text.isEmpty {
                texts.append((y: y, text: text))
            }
        }
        
        return texts
    }
    
    /// Gets the smallest font size used in the PDF, useful for determining scan height
    func getSmallestFontSize(from page: PDFPage) -> CGFloat {
        var fontSizes: Set<CGFloat> = []
        
        if let pageContent = page.attributedString {
            pageContent.enumerateAttribute(.font, in: NSRange(location: 0, length: pageContent.length)) { font, _, _ in
                if let ctFont = font as! CTFont? {
                    fontSizes.insert(CTFontGetSize(ctFont))
                }
            }
        }
        
        return fontSizes.min() ?? 5.0 // Default to 5pt if no fonts found
    }
}

File: android/app/src/main/java/com/farmaciasdeGuardia/services/pdfparsing/ColumnBasedPDFParser.kt
/*
 * Farmacias de Guardia - Segovia
 * Copyright (C) 2024 Bruno Follón
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package com.farmaciasdeGuardia.services.pdfparsing

import com.bfollon.farmaciasdeGuardia.util.DebugConfig
import org.apache.pdfbox.pdmodel.PDPage
import org.apache.pdfbox.text.PDFTextStripper
import org.apache.pdfbox.text.TextPosition
import java.io.StringWriter

/**
 * Base class for parsing PDFs that use a column-based layout.
 * Android equivalent of iOS ColumnBasedPDFParser.
 */
abstract class ColumnBasedPDFParser {
    
    /**
     * Represents a column of text in the PDF with its position and content
     */
    data class TextColumn(
        val x: Float,
        val width: Float,
        val texts: MutableList<Pair<Float, String>> = mutableListOf()
    )
    
    /**
     * Custom PDFTextStripper that captures text positions
     */
    private inner class PositionalTextStripper : PDFTextStripper() {
        val textBlocks = mutableListOf<TextBlock>()
        
        data class TextBlock(
            val x: Float,
            val y: Float,
            val width: Float,
            val height: Float,
            val text: String
        )
        
        override fun writeString(text: String, textPositions: List<TextPosition>) {
            if (text.trim().isEmpty()) return
            
            val firstPos = textPositions.firstOrNull() ?: return
            val lastPos = textPositions.lastOrNull() ?: firstPos
            
            textBlocks.add(
                TextBlock(
                    x = firstPos.x,
                    y = firstPos.y,
                    width = lastPos.x + lastPos.width - firstPos.x,
                    height = firstPos.height,
                    text = text.trim()
                )
            )
        }
    }
    
    /**
     * Extract text from a PDF page with positional information
     */
    protected fun extractTextWithPositions(page: PDPage): List<PositionalTextStripper.TextBlock> {
        return try {
            val stripper = PositionalTextStripper()
            stripper.startPage = 1
            stripper.endPage = 1
            
            // Process the page to extract positioned text
            val writer = StringWriter()
            stripper.writeText(page.document, writer)
            
            // Sort blocks by Y position (top to bottom), then X position (left to right)
            stripper.textBlocks.sortedWith(
                compareBy<PositionalTextStripper.TextBlock> { -it.y }.thenBy { it.x }
            )
        } catch (e: Exception) {
            DebugConfig.debugPrint("❌ Error extracting text positions: ${e.message}")
            emptyList()
        }
    }
    
    /**
     * Group text blocks into columns based on X position
     */
    protected fun groupIntoColumns(
        textBlocks: List<PositionalTextStripper.TextBlock>,
        columnThreshold: Float = 50f
    ): List<TextColumn> {
        if (textBlocks.isEmpty()) return emptyList()
        
        val columns = mutableListOf<TextColumn>()
        
        for (block in textBlocks) {
            // Find existing column or create new one
            val existingColumn = columns.find { column ->
                kotlin.math.abs(block.x - column.x) <= columnThreshold
            }
            
            if (existingColumn != null) {
                existingColumn.texts.add(Pair(block.y, block.text))
            } else {
                val newColumn = TextColumn(
                    x = block.x,
                    width = block.width
                )
                newColumn.texts.add(Pair(block.y, block.text))
                columns.add(newColumn)
            }
        }
        
        // Sort columns by X position and sort texts within each column by Y position
        return columns.sortedBy { it.x }.map { column ->
            column.copy(texts = column.texts.sortedBy { -it.first }.toMutableList())
        }
    }
    
    /**
     * Remove duplicate adjacent text blocks, keeping only the first occurrence
     */
    protected fun removeDuplicateAdjacent(blocks: List<Pair<Float, String>>): List<Pair<Float, String>> {
        return PDFParsingUtils.removeDuplicateAdjacent(blocks)
    }
    
    /**
     * Extract text from columns and flatten into lists
     */
    protected fun extractColumnTextFlattened(
        page: PDPage,
        expectedColumns: Int = 3
    ): Triple<List<String>, List<String>, List<String>> {
        val textBlocks = extractTextWithPositions(page)
        val columns = groupIntoColumns(textBlocks)
        
        DebugConfig.debugPrint("📄 Extracted ${columns.size} columns from PDF page")
        
        // Ensure we have at least the expected number of columns
        val paddedColumns = columns.toMutableList()
        while (paddedColumns.size < expectedColumns) {
            paddedColumns.add(TextColumn(0f, 0f))
        }
        
        // Extract text from each column
        val dates = paddedColumns.getOrNull(0)?.texts
            ?.map { it.second }
            ?.filter { it.isNotBlank() } ?: emptyList()
            
        val dayShifts = paddedColumns.getOrNull(1)?.texts
            ?.map { it.second }
            ?.filter { it.isNotBlank() } ?: emptyList()
            
        val nightShifts = paddedColumns.getOrNull(2)?.texts
            ?.map { it.second }
            ?.filter { it.isNotBlank() } ?: emptyList()
        
        DebugConfig.debugPrint("📄 Column data - Dates: ${dates.size}, Day shifts: ${dayShifts.size}, Night shifts: ${nightShifts.size}")
        
        return Triple(dates, dayShifts, nightShifts)
    }
}
